### PR TITLE
Faster Float64^Float64

### PR DESCRIPTION
**Pull request by oscardssmith**
I've finally managed to get an accurate version of `x^y` working for Float64 inputs. I think it's still missing some NaN checks and similar, but the core hard part is now done. Massive thanks to @chriselrod for hanging out with me for the 3.5 hours it took to debug the compensated arithmetic in the extended precision log (and giving me the SLEEFPirates reference implementation). Initial benchmarks:
Old:
```
@btime ^($(Ref(3.14))[], $(Ref((2.7))[]))
  66.043 ns (0 allocations: 0 bytes)
```
New:
```
@btime ^($(Ref(3.14))[], $(Ref((2.7))[]))
  24.237 ns (0 allocations: 0 bytes)
```

Theoretically, this should be accurate to 1 ULP, but I haven't done extensive testing yet.

### CODE REVIEW

**Comment by chriselrod** [comment]
It'd be great to get LICM on our Julia implementations of "pure" functions like `^` and `exp`, it makes a huge difference in examples like [this](https://discourse.julialang.org/t/defining-function-as-scalar-vs-fusing-with-two-refs-significant-speed-difference/68273/2?u=elrod).
Because `^` here is using `@llvm.pow`, LLVM will licm it. It won't anymore once we switch to a Julia implementation.

**Comment by oscardssmith** [comment]
The other thing that would be really nice is if the compiler was smart enough to precompute `_log_ext(x)` if a user writes something like `3.1^[4.0,6.0,1.0]`

**Comment by chriselrod** [comment]
How many people would complain if we `@inline` everything, or at least the `_log_ext(x)`?

**Comment by oscardssmith** [comment]
We can't inline the whole thing, but we could inline `_log_ext` if we don't inline `pow`.

File: base/special/log.jl [hunk 1]
```diff
@@ -409,3 +409,56 @@ function log1p(x::Float32)
...
+    c1lo = 3.80554962542412056336616e-17
+    estr =  evalpoly(x_hi, (c2, c3, c4, c5, c6, c7, c8, c9, c10))
+    res_hi = estr*x_hi
+    res_lo = fma(x_lo, estr, fma(estr, x_hi, -res_hi))
+    ans_hi = c1hi + res_hi
+    ans_lo = (c1hi - ans_hi) + res_hi + (res_lo + c1lo)
+    return ans_hi, ans_lo
+end
+
+# Log implimentation that returns 2 numbers which sum to give true value with about 68 bits of precision
```
Comment (KristofferC) [inline base/special/log.jl:433]:
```suggestion
# Log implementation that returns 2 numbers which sum to give true value with about 68 bits of precision
```

I don't really understand "which sum to give true value". Also, is SleefPirates.jl the original source of this code or was it adapted from somewhere else originally?

Comment (oscardssmith) [reply base/special/log.jl]: what I mean by this is that it outputs `yhi,ylo` such that `log(big(x))-yh-ylo <2^-68`, but the output is not normalized (`ylo !< eps(yhi)`). Basically for true double-double implementations you would typically guarantee that all the bits in `yhi` are correct, but I'm not doing that because it's not necessary for `pow`

Comment (oscardssmith) [reply base/special/log.jl]: @chriselrod where did the SLEEFPirates code come from again?

Comment (chriselrod) [reply base/special/log.jl]: SLEEF: https://github.com/shibatch/sleef
@musm 's port: https://github.com/musm/SLEEF.jl
SLEEFPirates is a fork of the port.

Comment (chriselrod) [reply base/special/log.jl]: Probably?
The SLEEF.jl license mentions it's a port of SLEEF and includes a link, but has a different license and doesn't include a copy of SLEEF's.

I don't care about SLEEFPirates/its license, so probably just SLEEF.jl's is fine.

**Comment by chriselrod** [comment]
> We can't inline the whole thing, but we could inline `_log_ext` if we don't inline `pow`.

Actually, given the addition of callsite inlining, I think these functions should take the approach of not being inlined, but having all code within them inlined so that someone can do `@inline a ^ b` to get a fully inlined function.
Then it's also worth testing to ensure that the implementations are SIMD-able, when possible.

**Comment by oscardssmith** [comment]
After a long period, this is finally ready to go! My plan is to merge this and https://github.com/JuliaLang/julia/pull/40620 after https://github.com/JuliaLang/julia/pull/42031 is merged so that we don't have a regression for integer powers.

**Comment by oscardssmith** [comment]
Merging tomorrow Sans objection.

**Comment by barucden** [comment]
I am sorry -- is it intentional to remove the `rem` implementation? It seems that this PR, besides improving `^`, basically reverts #42380.

**Comment by oscardssmith** [comment]
Good catch. I got sloppy with the rebase.

**Comment by oscardssmith** [comment]
@KristofferC can you run PackageEval on this?

**Comment by KristofferC** [comment]
Sure, but I think you can do it too.

@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/42271#issuecomment-938385992) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/da8da62_vs_55808a5/report.md).

**Comment by oscardssmith** [comment]
TLDR is there are a few packages that break, but only ones that do are testing with explicit floating point values, and are off by a few ULPs (and they are testing more complicated functions), so we appear to be good to merge.

**Comment by oscardssmith** [comment]
Merging on Tuesday if there aren't objections.

**Comment by Keno** [comment]
This blew up compile times in MTK. cc @JeffBezanson

File: base/math.jl [hunk 2]
```diff
@@ -947,11 +947,15 @@ end
 @inline function ^(x::Float64, y::Float64)
```
Comment (Keno) [inline base/math.jl:947]: This `@inline` seems potentially problematic now that this function is no longer small.
